lib/streamlit/runtime/state/query_params.py

```diff
@@ -0,0 +1,126 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass, field
+from typing import Any, Dict, Iterator, List, MutableMapping, Union
+
+from streamlit.proto.ForwardMsg_pb2 import ForwardMsg
+
+
+@dataclass
+class QueryParams(MutableMapping[str, Any]):
+    """A lightweight wrapper of a dict that sends forwardMsgs when state changes.
+    It stores str keys with str and List[str] values.
+    """
+
+    _query_params: Dict[str, Union[List[str], str]] = field(default_factory=dict)
+
+    def __init__(self):
+        # avoid using ._query_params as that will use __setattr__,
+        # which itself relies on `_query_params` being defined
+        self.__dict__["_query_params"] = {}
+
+    def __iter__(self) -> Iterator[Any]:
+        return iter(self._query_params.keys())
+
+    def __getitem__(self, key: str) -> str:
+        return self._getitem(key)
+
+    def _getitem(self, key: str) -> str:
+        try:
+            value = self._query_params[key]
+            if isinstance(value, list):
+                if len(value) == 0:
+                    return ""
+                else:
+                    return value[-1]
+            return value
+        except:
+            raise KeyError(_missing_key_error_message_query_params(key))
+
+    def _setitem(self, key: str, value: Union[str, List[str]]) -> None:
+        if isinstance(value, list):
+            self._query_params[key] = [str(item) for item in value]
+        else:
+            self._query_params[key] = str(value)
+        self._send_query_param_msg()
+
+    def __setitem__(self, key: str, value: str) -> None:
+        self._setitem(key, value)
+
+    def get_all(self, key: str) -> List[str]:
+        try:
+            if key not in self._query_params:
+                return []
+            query_params = self._query_params[key]
+            return query_params if isinstance(query_params, list) else [query_params]
+        except KeyError:
+            raise KeyError(_missing_key_error_message_query_params(key))
+
+    def __getattr__(self, key: str) -> str:
+        try:
+            return self._getitem(key)
+        except KeyError:
+            raise AttributeError(_missing_key_error_message_query_params(key))
+
+    def __setattr__(self, key: str, value: str) -> None:
+        self._setitem(key, value)
+
+    def __delattr__(self, key: str) -> None:
+        try:
+            self._delitem(key)
+        except KeyError:
+            raise AttributeError(_missing_key_error_message_query_params(key))
+
+    def _delitem(self, key):
+        if key in self._query_params:
+            del self._query_params[key]
+            self._send_query_param_msg()
+        else:
+            raise KeyError(_missing_key_error_message_query_params(key))
+
+    def __contains__(self, key: str) -> bool:  # type: ignore[override]
```

Comment:
It felt appropriate to have this be a str and override it but happy to change it back to the original python function type...

Reply:
Is the override needed so that `key` can be typed as `str` and not `Any`? That's the typing we want here and not the current one. There's nothing wrong with checking if a non-`str` is contained within a mapping with keys of type `str`. We just know that the return value of that should always be `False`